tests/system/src/test/java/test/robot/javafx/scene/TextSurrogatePairInsertionIndexTest.java

```diff
@@ -0,0 +1,125 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package test.robot.javafx.scene;
+
+import java.util.Random;
+import java.util.concurrent.CountDownLatch;
+import javafx.application.Application;
+import javafx.application.Platform;
+import javafx.geometry.Bounds;
+import javafx.geometry.Point2D;
+import javafx.scene.Scene;
+import javafx.scene.layout.BorderPane;
+import javafx.scene.text.Font;
+import javafx.scene.text.HitInfo;
+import javafx.scene.text.Text;
+import javafx.stage.Stage;
+import javafx.stage.StageStyle;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import test.util.Util;
+
+/*
+ * Verifies insertion index computation in Text in the presence of surrogate pairs.
+ */
+public class TextSurrogatePairInsertionIndexTest {
+    private static final int FONT_SIZE = 36;
+    private static final int WIDTH = 200;
+    private static final int HEIGHT = FONT_SIZE * 7;
+
+    private static CountDownLatch startupLatch = new CountDownLatch(1);
+    private static Random random;
+    private static Text control;
+    private static Stage stage;
+    private static Scene scene;
+    private static BorderPane content;
+
+    private void create(String text) {
+        control = new Text(text);
+        control.setFont(new Font(FONT_SIZE));
+        content.setTop(control);
+        control.wrappingWidthProperty().bind(content.widthProperty());
+    }
+
+    /**
+     * Test tests HitInfo invariants randomly picking a point within the Text instance.
+     */
+    @Test
+    public void testHitInfo() {
+        Util.runAndWait(() -> {
+            create("[This is text 😀😃😄\n\n😁😆 💙🦋🏁🔥]");
+        });
+        Util.waitForIdle(scene);
+
+        Util.runAndWait(() -> {
+            Bounds b = control.getLayoutBounds();
+            Assert.assertTrue(b.getWidth() > 10);
+            Assert.assertTrue(b.getHeight() > 10);
+
+            int max = 1_000_000;
+            for (int i = 0; i < max; i++) {
+                double x = random.nextDouble() * b.getWidth();
+                double y = random.nextDouble() * b.getHeight();
+                HitInfo h = control.hitTest(new Point2D(x, y));
+
+                Assert.assertTrue(h.getInsertionIndex() >= 0);
+                String s = h.toString();
+                Assert.assertTrue(s != null);
+            }
+        });
+    }
+
+    @BeforeClass
+    public static void initFX() {
+        long seed = new Random().nextLong();
+        System.out.println("seed=" + seed);
```

Comment:
Do we need this print statement?

Reply:
yes.  I've added a brief comment, but let me explain in more detail:

these tests use random patterns to avoid hitting the same values each time.  for example, in TextFlowSurrogatePairInsertionIndexTest many tests incremented the x coordinate by 5 prior to this change, but what if we had an error condition happening with x = 21?.  the old test never hit that value, but the new one might (at some point).  so we effectively extending the test coverage every time the test runs.

and, if we ever encounter a failure, it will be possible to reproduce the exact failure scenario by setting the same seed as found in the failed log.

Reply:
Thanks for the details @andy-goryachev-oracle